Defining Serverless Workflow Notation

tsurdilo opened the pull request:
In addition to facilitate portability of serverless workflow model (json) it is also important that we facilitate portability of the workflow notation between tooling. 

This PR adds the portable workflow notation which is to be used by tooling. 

I have created all images/icons myself so there are no copyright issues and  am contributing everything to cncf wg (or whoever this needs to be contributed to). 

You can see what this update looks like on one single page here:  https://github.com/tsurdilo/wg-serverless/blob/workflownotation/workflow/spec/spec-notation.md

tsurdilo commented:
depends on https://github.com/cncf/wg-serverless/pull/82

cathyhongzhang commented on workflow/spec/spec-notation.md:
```diff
@@ -0,0 +1,220 @@
+| Trigger Event |<img src="media/notation/trigger-notation.png" width="135px" height="110px"/> | Trigger Events have to be placed inside the <br> "Events" section of the workflow. |
+| Event State |<img src="media/notation/event-state-notation.png" width="135px" height="110px"/> | Includes actions. <br> Can be expaded if the "+" icon is clicked. | 
+| Operation State |<img src="media/notation/operation-state-notation.png" width="135px" height="110px"/> | Includes actions. <br> Can be expaded if the "+" icon is clicked.  | 
+| Switch State |<img src="media/notation/switch-state-notation.png" width="135px" height="110px"/> | Includes choices. <br> Can be expaded if the "+" icon is clicked.  | 
```
I am not sure if it is only me who has the problem. The image does not show properly (no + icon shows up). Could you fix this so that I can review it? Thanks.

tsurdilo replied:
You mean in the github "files changes" view? There probably wont show binary files. You can either fork my local repo (workflownotation branch) and look at it there (https://github.com/tsurdilo/wg-serverless.git) or as mentioned in the initial comment
view directly here:  https://github.com/tsurdilo/wg-serverless/blob/workflownotation/workflow/spec/spec-notation.md
Hope this helps.

ruromero replied:
I can see the images. However I would like to make some suggestions. For switch I think something like this https://www.iconfinder.com/icons/4404399/choice_decision_option_way_icon but without the ticks

cathyhongzhang replied:
I would also like to make some suggestions on the trigger event : 
https://icon-library.net/icon/trigger-icon-12.html

switch state: https://www.iconfinder.com/icons/3003425/chance_choice_chosen_decision_quesion_risk_icon

I would also like to make some suggestions. For trigger event, something like this https://icon-library.net/icon/trigger-icon-12.html. Event state is "trigger event + operation state". For switch, something like https://www.iconfinder.com/icons/3003425/chance_choice_chosen_decision_quesion_risk_icon
since a switch state could have multiple branches. Ruben's suggestion is a good fit for single choice.

tsurdilo replied:
@cathyhongzhang sure I will try to incorporate all the suggestions. Those are good icons to use I think as well.

Would you look at https://github.com/tsurdilo/wg-serverless/blob/workflownotation/workflow/spec/spec-notation.md as I recently changed alot of the icons that had originally. Let me know what you think

tsurdilo commented:
ill fix merge conflicts once a decision on this has been made (before merging)

ruromero commented on workflow/spec/spec.md:
```diff
@@ -1040,4 +1041,10 @@ Each Filter has three kinds of path filters
+
+## Notation
+
+In addition to facilitating portability of the serverless workflow model, it is as important
```
```suggestion
In addition to facilitating portability of the serverless workflow model, it is also important
```

ruromero commented on workflow/spec/spec-notation.md:
```diff
@@ -0,0 +1,220 @@
+| Workflow |<img src="media/notation/workflow-notation.png" width="135px" height="110px"/> | Workflow consists of two sections, Events and States. <br>  Events section can only include Trigger Events. <br> State Section incldues all workflow states. <br> If no trigger events are defined <br> the Events section does not have to be displayed.  |
+| Trigger Event |<img src="media/notation/trigger-notation.png" width="135px" height="110px"/> | Trigger Events have to be placed inside the <br> "Events" section of the workflow. |
+| Event State |<img src="media/notation/event-state-notation.png" width="135px" height="110px"/> | Includes actions. <br> Can be expaded if the "+" icon is clicked. | 
+| Operation State |<img src="media/notation/operation-state-notation.png" width="135px" height="110px"/> | Includes actions. <br> Can be expaded if the "+" icon is clicked.  | 
```
Can I suggest something like a gear? This icon can be seen as a loop or something recurrent.

tsurdilo commented:
@ruromero updated all the images please have another look and let me know what you think.
updated all the examples (removed end state..added end states section for each state if "end": true).
rebased against current master

cathyhongzhang commented on workflow/spec/spec-notation.md:
```diff
@@ -0,0 +1,226 @@
+#### Labels
+
+Labels can be added to all elements of the serverless workflow notation. 
+They should be localged cenetered and underneith the element icon. 
```
please fix typos

tsurdilo replied:
fixed. thanks.

cathyhongzhang commented on workflow/spec/spec.md:
```diff
@@ -988,6 +989,11 @@ Each Filter has three kinds of path filters
+the serverless workflow notation. Details on this can be found [here](spec-notation.md)
+
 ## Reference
 You can find a list of other languages, technologies and specifications related to workflows [here](references.md)
```
Thanks for all these notation suggestions. It may be good to have a f2f discussion on the notation in the kubecon. Hopefully every notation we put in the spec is self-explanatory.

tsurdilo commented:
Putting this pr "On Hold" as we are focusing on the specification core atm. Will revisit and ask for community feedback in future.
